bundles/org.smarthomej.transform.basicprofiles/src/main/java/org/smarthomej/transform/basicprofiles/internal/profiles/RoundStateProfile.java

```diff
@@ -0,0 +1,136 @@
+import org.openhab.core.thing.profiles.ProfileCallback;
+import org.openhab.core.thing.profiles.ProfileContext;
+import org.openhab.core.thing.profiles.ProfileTypeUID;
+import org.openhab.core.thing.profiles.StateProfile;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.openhab.core.types.Type;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.smarthomej.transform.basicprofiles.internal.config.RoundStateProfileConfig;
+
+/**
+ * Applies rounding with the specified scale and the rounding mode to a {@link QuantityType} or {@link DecimalType}
+ * state. Default rounding mode is {@link RoundingMode#HALF_UP}.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class RoundStateProfile implements StateProfile {
+
+    private final Logger logger = LoggerFactory.getLogger(RoundStateProfile.class);
+
+    private final ProfileCallback callback;
+
+    final int scale;
+    final RoundingMode roundingMode;
+
+    public RoundStateProfile(ProfileCallback callback, ProfileContext context) {
+        this.callback = callback;
+        RoundStateProfileConfig config = context.getConfiguration().as(RoundStateProfileConfig.class);
+        logger.debug("Configuring profile with parameters: [{scale='{}', mode='{}']", config.scale, config.mode);
+
+        int localScale = 0;
+        // if (config.scale instanceof String) {
+        // try {
+        // scale = Integer.valueOf((String) config.scale);
+        // } catch (NumberFormatException e) {
+        // logger.error("Cannot convert value '{}' of parameter 'scale' into a valid integer.", config.scale);
+        // }
+        // } else
+        Integer configScale = config.scale;
+        if (configScale instanceof Number) {
+            localScale = ((Number) configScale).intValue();
+        } else {
+            logger.error("Parameter 'scale' is not of type String or Number.");
+        }
+
+        if (localScale < 0) {
+            logger.warn("Parameter 'scale' has to be a non-negative integer. Ignoring it.");
+            localScale = 0;
+        }
+
+        RoundingMode localRoundingMode = RoundingMode.HALF_UP;
+        if (config.mode instanceof String) {
+            try {
+                localRoundingMode = RoundingMode.valueOf(config.mode);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Parameter 'mode' is not a supported rounding mode: '{}'. Using default.", config.mode);
+            }
+        } else {
+            logger.error("Parameter 'mode' is not of type String.");
+        }
+
+        this.scale = localScale;
+        this.roundingMode = localRoundingMode;
+    }
+
+    @Override
+    public ProfileTypeUID getProfileTypeUID() {
+        return ROUND_UID;
+    }
+
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        callback.sendUpdate((State) applyRound(state));
```

Comment:
Is that correct? I thought that `sendUpdate` updates the item's state. But shouldn't this result in a state that is processed by the handler?

Reply:
No. Round is a one-way Profile. It should be applied only in updates / commands from handler towards Item.

Reply:
@cweitkamp I would like to come back to this comment, as the current implementation causes issues with latest core changes. I also came across this when using the invert/negate profile.

As a result of this line we end up in an endless loop: 

1. The item updates its state to `123.12` and sends an `ItemStateUpdatedEvent` with a state of `123.12`.
2. The `CommunicationManager` forwards this state update to the and according to its configuration the profile rounds that to `123` and forwards that to the callback's `sendUpdate` method.
3. The callback issues an `ItemStateEvent` which is received by the item, which sets its state to `123`. After that, it sends an `ItemStateUpdatedEvent` with a state of `123`.
4. The `CommunicationManager` receives that, ... (start again at 2).

the same issue arises with the invert profile, because `CLOSED` is inverted to `OPEN` and after updating the item state to `OPEN` it is again inverted to `CLOSED`, resulting in the same update loop.

We can't prevent that by checking event `source`, because the `ItemUpdater` (that receives the `ItemStateEvent`) calls `GenericItem.setState` (which creates the `ItemStateUpdatedEvent/ItemStateChangedEvent`) and the `source` is lost (otherwise we could have set the `source` of the `ItemStateUpdatedEvent` to the same value as the `ItemStateEvent`.

IMO a profile should only change values that are transferred from the handler to the item (or from the item to the handler), not interact with the item state itself.

Reply:
And, obviously, it is right now a blocking issue to migrate to OH 4 with M2 or newer snapshots, so maybe this can be resolved somehow.

Reply:
Your explanation sounds reasonable. I won't argue against it. Feel free to change/remove this part of code.

Reply:
After upgrading to latest OH milestone I am suffering from this myself. I submitted https://github.com/smarthomej/addons/pull/478 to fix this.